proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -507,39 +507,1153 @@ locale delete_one = delete_one_conc + delete_one_abs +
+    apply (simp only: dom_fun_upd2 simp_thms)
+    apply (elim conjE)
+    apply (frule bspec, erule domI)
+    apply (frule_tac x=nrp in bspec, erule domI)
+    apply (rule ballI, drule(1) bspec)
+    apply (drule domD)
+    apply (prop_tac "scp \<noteq> nrp", clarsimp)
+    apply (clarsimp simp: project_inject split: if_split_asm kernel_object.split_asm)
+    apply (rule conjI; clarsimp)
+     apply (clarsimp simp: sc_relation_def)
+    apply (rename_tac bb aa ba)
+    apply (drule_tac x="(aa, ba)" in bspec, simp)
+    apply (clarsimp simp: obj_at_def is_reply)
+    apply (frule_tac ko'="kernel_object.Reply reply" and x'=nrp in obj_relation_cut_same_type)
+       apply simp+
+    apply (clarsimp simp: reply_relation_def)
+   apply (rule conjI)
+    (* sc_replies_relation *)
+    apply (clarsimp simp: projectKO_opt_sc opt_map_left_Some)
+    apply (frule (2) StateRelation.sc_replies_relation_scReply)
+    apply (clarsimp simp: sc_replies_relation_def sc_replies_of_scs_def map_project_def scs_of_kh_def)
+    apply (drule_tac x=p in spec)
+    apply (intro conjI impI allI)
+     (* p =  scp *)
+     apply (clarsimp simp: opt_map_left_Some sc_of_def)
+     apply (prop_tac "replyPrevs_of b nrp = Some rp")
+      apply (simp add: sym_refs_replyNext_replyPrev_sym[symmetric])
+      apply (clarsimp simp: opt_map_left_Some)
+     apply (prop_tac "scReply sc' \<noteq> Some rp")
+      apply (drule heap_path_head; clarsimp)
+     apply (frule (4) heap_ls_next_takeWhile_append_sym[OF _ _ _ _ reply_sym_heap_Prev_Next])
+     apply simp
+     apply (rule heap_path_heap_upd_not_in[rotated])
+      using set_takeWhileD apply (metis (full_types))
+     apply (erule heap_path_takeWhile_lookup_next)
+     apply (prop_tac "nrp \<in> set (takeWhile ((\<noteq>) rp) (sc_replies sc))")
+      apply clarsimp
+     using set_takeWhileD apply metis
+    (* p \<noteq> scp *)
+    apply (rule heap_path_heap_upd_not_in)
+     apply clarsimp
+    apply (clarsimp simp: sc_of_def opt_map_Some)
+    apply (rename_tac p sc2 ko)
+    apply (case_tac ko; simp)
+    apply (clarsimp simp: opt_map_left_Some sc_of_def)
+    apply (prop_tac "replyPrevs_of b nrp = Some rp")
+     apply (simp add: sym_refs_replyNext_replyPrev_sym[symmetric])
+     apply (clarsimp simp: opt_map_left_Some)
+    apply (frule (2) heap_ls_next_in_list)
+    apply (simp add: sc_with_reply_def the_pred_option_def
+              split: if_split_asm)
+    apply blast
+   apply (rule conjI)
+    apply (clarsimp simp add: ghost_relation_def)
+    apply (erule_tac x=scp in allE)+
+    apply (clarsimp simp: obj_at_def a_type_def
+                   split: Structures_A.kernel_object.splits if_split_asm)
+   apply (clarsimp simp: caps_of_state_after_update cte_wp_at_after_update
+                         swp_def fun_upd_def obj_at_simps)
+   done
+qed
+
+end
+
+(* end : related corres rules *)
+
+(* FIXME RT: the original in Invariants_AI; do we want valid_objs or valid_tcbs? *)
```

Comment:
Yeah I'm not sure the `valid_tcbs/valid_tcbs'` experiment worked out... I think it's been more trouble than it's worth.

Maybe replace `clarsimp simp add:` with just `clarsimp simp:`